Reflector watchHandler: make 'The resourceVersion for ... watch is too old' log Info not Warning

sallyom opened the pull request:
**What type of PR is this?**
logging Warning to Info - not sure what /kind

**What this PR does / why we need it**:
This warning comes from Reflector watchHandler, from the apiserver error that indicates a watch was restarted.  This happens when etcd drops the connection and resources are relisted.  This informs the user that the watchers are operating properly, so should be logged as Info rather than Warning.

```release-note
Reflector watchHandler Warning log 'The resourceVersion for the provided watch is too old.' is now logged as Info. 
```

enj commented:
/kind cleanup
/ok-to-test
/lgtm

enj commented:
/priority important-soon

sallyom commented:
this is my first k8s PR, I think this should be noted :tada:

enj commented:
> this is my first k8s PR, I think this should be noted :tada:

The bot totally did notice: https://github.com/kubernetes/kubernetes/pull/78991#issuecomment-501824181

/lgtm

sallyom commented:
/assign @sttts

liggitt commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -298,7 +300,12 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
+			switch err {
+			case errorStopRequested:
+				// no extra log message
+			case etcdrpc.ErrCompacted:
```
this layer doesn't currently have a dependency on or knowledge of etcd... can we propagate this in a way that doesn't make us deep compare to specific etcd errors?

liggitt replied:
actually... an API client isn't going to get raw etcd errors... it will get an API error

sallyom replied:
lemme rework that

sallyom replied:
_sigh_ ok fixed that

fedebongio commented:
/assign @jingyih

fedebongio commented:
/assign @roycaihw

sallyom committed:
Modify klog Warning 'The resourceVersion for the provided watch is too old' to Info.

This warning comes from Reflector watchHandler, from the apiserver error that
indicates a watch was restarted.  This happens when etcd drops the connection
and resources are relisted.  This informs the user that the watchers are
operating properly, so should be logged as Info rather than Warning.

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -78,6 +78,7 @@ var (
 	// We try to spread the load on apiserver by setting timeouts for
 	// watch requests - it is random in [minWatchTimeout, 2*minWatchTimeout].
 	minWatchTimeout = 5 * time.Minute
+	ErrStaleWatch   = apierrs.NewResourceExpired("The resourceVersion for the provided watch is too old.")
```
https://github.com/kubernetes/kubernetes/blob/b3875556b0edf3b5eaea32c69678edcf4117d316/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L467

consider reuse the error for the watch-cache?

sttts replied:
I fear we cannot access that here from client-go.

sttts commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -298,7 +299,12 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
+			case errorStopRequested:
+				// no extra log message
+			case ErrStaleWatch:
+				klog.V(1).Infof("%s: watch of %v ended with: %v", r.name, r.expectedType, err)
```
would even go higher with the verbosity level. 4 maybe. So it is hidden normally.

sallyom replied:
done, made it a 4

sttts commented on staging/src/k8s.io/client-go/tools/cache/reflector.go:
```diff
@@ -298,7 +299,12 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
+			switch err {
+			case errorStopRequested:
+				// no extra log message
+			case ErrStaleWatch:
```
this compares the human readable message (next to anything else of the error). If you look what the error actually is:
```
	return &StatusError{metav1.Status{
		Status:  metav1.StatusFailure,
		Code:    http.StatusGone,
		Reason:  metav1.StatusReasonExpired,
		Message: message,
	}}
```

we should probably just compare the reason (meant to be machine readable). There is even a func `IsResourceExpired` for that. Then you also don't have to export the `ErrStaleWatch` from here, and we avoid coupling of our client library with some low level storage library for etcd.

sallyom replied:
thanks! I did not look that far, now I know :)

sttts commented:
/lgtm
/approve
